shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h

```diff
@@ -0,0 +1,148 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_COMMON_CPP_CLIENT_WRAPPER_INCLUDE_FLUTTER_EVENT_CHANNEL_H_
+#define FLUTTER_SHELL_PLATFORM_COMMON_CPP_CLIENT_WRAPPER_INCLUDE_FLUTTER_EVENT_CHANNEL_H_
+
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include "binary_messenger.h"
+#include "engine_method_result.h"
+#include "event_sink.h"
+#include "event_stream_handler.h"
+
+namespace flutter {
+
+// A named channel for communicating with the Flutter application using
+// asynchronous event streams. Incoming requests for event stream setup are
+// decoded from binary on receipt, and C++ responses and events are encoded into
+// binary before being transmitted back to Flutter. The MethodCodec used must be
+// compatible with the one used by the Flutter application. This can be achieved
+// by creating an EventChannel
+// ("https://docs.flutter.io/flutter/services/EventChannel-class.html")
+// counterpart of this channel on the Dart side.
+// The C++ type of stream configuration arguments, events, and error details are
+// templated, but only values supported by the specified MethodCodec can be
+// used.
+template <typename T>
+class EventChannel {
+ public:
+  // Creates an instance that sends and receives event handler on the channel
+  // named |name|, encoded with |codec| and dispatched via |messenger|.
+  EventChannel(BinaryMessenger* messenger,
+               const std::string& name,
+               const MethodCodec<T>* codec)
+      : messenger_(messenger), name_(name), codec_(codec) {}
+  ~EventChannel() = default;
+
+  // Prevent copying.
+  EventChannel(EventChannel const&) = delete;
+  EventChannel& operator=(EventChannel const&) = delete;
+
+  // Registers a stream handler on this channel.
+  // If no handler has been registered, any incoming stream setup requests will
+  // be handled silently by providing an empty stream.
+  void SetStreamHandler(StreamHandler<T>* handler) {
+    if (!handler) {
+      messenger_->SetMessageHandler(name_, nullptr);
+      is_listened_ = false;
+      return;
+    }
+
+    const MethodCodec<T>* codec = codec_;
+    const std::string channel_name = name_;
+    const BinaryMessenger* messenger = messenger_;
+    bool& is_listend = is_listened_;
+    BinaryMessageHandler binary_handler = [handler, codec, channel_name,
+                                           messenger, &is_listend](
+                                              const uint8_t* message,
+                                              const size_t message_size,
+                                              BinaryReply reply) {
+      constexpr char kOnListenMethod[] = "listen";
+      constexpr char kOnCancelMethod[] = "cancel";
+
+      std::shared_ptr<MethodCall<T>> method_call =
```

Comment:
I don't see any reason for this to be a `shared_ptr` rather than a `unique_ptr`.

Reply:
Sorry, this is my mistake.  `unique_ptr` is correct.